File: backend/internal/models/analytics.go
package models

import "time"

// TimeRange 時間範圍類型
type TimeRange string

const (
	TimeRangeWeek    TimeRange = "week"    // 本週
	TimeRangeMonth   TimeRange = "month"   // 本月
	TimeRangeQuarter TimeRange = "quarter" // 本季
	TimeRangeYear    TimeRange = "year"    // 本年
	TimeRangeAll     TimeRange = "all"     // 全部
)

// Validate 驗證時間範圍是否有效
func (tr TimeRange) Validate() bool {
	switch tr {
	case TimeRangeWeek, TimeRangeMonth, TimeRangeQuarter, TimeRangeYear, TimeRangeAll:
		return true
	default:
		return false
	}
}

// GetDateRange 根據時間範圍取得起始和結束日期
func (tr TimeRange) GetDateRange() (startDate, endDate time.Time) {
	now := time.Now()
	endDate = now

	switch tr {
	case TimeRangeWeek:
		// 本週：從週一開始
		weekday := int(now.Weekday())
		if weekday == 0 {
			weekday = 7 // 週日視為第 7 天
		}
		startDate = now.AddDate(0, 0, -(weekday - 1))
	case TimeRangeMonth:
		// 本月：從 1 號開始
		startDate = time.Date(now.Year(), now.Month(), 1, 0, 0, 0, 0, now.Location())
	case TimeRangeQuarter:
		// 本季：從季度第一個月的 1 號開始
		month := now.Month()
		quarterStartMonth := ((month-1)/3)*3 + 1
		startDate = time.Date(now.Year(), quarterStartMonth, 1, 0, 0, 0, 0, now.Location())
	case TimeRangeYear:
		// 本年：從 1/1 開始
		startDate = time.Date(now.Year(), 1, 1, 0, 0, 0, 0, now.Location())
	case TimeRangeAll:
		// 全部：從很久以前開始（2000-01-01）
		startDate = time.Date(2000, 1, 1, 0, 0, 0, 0, now.Location())
	}

	return startDate, endDate
}

// AnalyticsSummary 分析摘要資料
type AnalyticsSummary struct {
	TotalRealizedPL    float64 `json:"total_realized_pl"`     // 總已實現損益
	TotalRealizedPLPct float64 `json:"total_realized_pl_pct"` // 總已實現損益百分比
	TotalCostBasis     float64 `json:"total_cost_basis"`      // 總成本基礎
	TotalSellAmount    float64 `json:"total_sell_amount"`     // 總賣出金額
	TotalSellFee       float64 `json:"total_sell_fee"`        // 總賣出手續費
	TransactionCount   int     `json:"transaction_count"`     // 交易筆數
	Currency           string  `json:"currency"`              // 幣別
	TimeRange          string  `json:"time_range"`            // 時間範圍
	StartDate          string  `json:"start_date"`            // 起始日期
	EndDate            string  `json:"end_date"`              // 結束日期
}

// PerformanceData 績效資料（按資產類型）
type PerformanceData struct {
	AssetType       AssetType `json:"asset_type"`        // 資產類型
	Name            string    `json:"name"`              // 資產類型名稱
	RealizedPL      float64   `json:"realized_pl"`       // 已實現損益
	RealizedPLPct   float64   `json:"realized_pl_pct"`   // 已實現損益百分比
	CostBasis       float64   `json:"cost_basis"`        // 成本基礎
	SellAmount      float64   `json:"sell_amount"`       // 賣出金額
	TransactionCount int      `json:"transaction_count"` // 交易筆數
}

// TopAsset 最佳/最差表現資產
type TopAsset struct {
	Symbol        string    `json:"symbol"`          // 標的代碼
	Name          string    `json:"name"`            // 標的名稱
	AssetType     AssetType `json:"asset_type"`      // 資產類型
	RealizedPL    float64   `json:"realized_pl"`     // 已實現損益
	RealizedPLPct float64   `json:"realized_pl_pct"` // 已實現損益百分比
	CostBasis     float64   `json:"cost_basis"`      // 成本基礎
	SellAmount    float64   `json:"sell_amount"`     // 賣出金額
}

// AssetTypeNameMap 資產類型名稱對應
var AssetTypeNameMap = map[AssetType]string{
	AssetTypeTWStock: "台股",
	AssetTypeUSStock: "美股",
	AssetTypeCrypto:  "加密貨幣",
}

// GetAssetTypeName 取得資產類型的顯示名稱
func GetAssetTypeName(assetType AssetType) string {
	if name, ok := AssetTypeNameMap[assetType]; ok {
		return name
	}
	return string(assetType)
}

// ==================== 未實現損益分析 ====================

// UnrealizedSummary 未實現損益摘要
type UnrealizedSummary struct {
	TotalCost          float64 `json:"total_cost"`           // 總成本
	TotalMarketValue   float64 `json:"total_market_value"`   // 總市值
	TotalUnrealizedPL  float64 `json:"total_unrealized_pl"`  // 總未實現損益
	TotalUnrealizedPct float64 `json:"total_unrealized_pct"` // 總未實現報酬率
	HoldingCount       int     `json:"holding_count"`        // 持倉數量
	Currency           string  `json:"currency"`             // 幣別
}

// UnrealizedPerformance 各資產類型未實現績效
type UnrealizedPerformance struct {
	AssetType     AssetType `json:"asset_type"`     // 資產類型
	Name          string    `json:"name"`           // 資產類型名稱
	Cost          float64   `json:"cost"`           // 成本
	MarketValue   float64   `json:"market_value"`   // 市值
	UnrealizedPL  float64   `json:"unrealized_pl"`  // 未實現損益
	UnrealizedPct float64   `json:"unrealized_pct"` // 未實現報酬率
	HoldingCount  int       `json:"holding_count"`  // 持倉數量
}

// UnrealizedTopAsset Top 未實現損益資產
type UnrealizedTopAsset struct {
	Symbol        string    `json:"symbol"`         // 標的代碼
	Name          string    `json:"name"`           // 標的名稱
	AssetType     AssetType `json:"asset_type"`     // 資產類型
	Quantity      float64   `json:"quantity"`       // 持有數量
	AvgCost       float64   `json:"avg_cost"`       // 平均成本
	CurrentPrice  float64   `json:"current_price"`  // 當前價格
	Cost          float64   `json:"cost"`           // 總成本
	MarketValue   float64   `json:"market_value"`   // 市值
	UnrealizedPL  float64   `json:"unrealized_pl"`  // 未實現損益
	UnrealizedPct float64   `json:"unrealized_pct"` // 未實現報酬率
}

File: frontend/src/types/analytics.ts
import { AssetType } from "./transaction";

// ==================== 列舉型別 ====================

/**
 * 時間範圍
 */
export const TimeRange = {
  WEEK: "week",
  MONTH: "month",
  QUARTER: "quarter",
  YEAR: "year",
  ALL: "all",
} as const;

export type TimeRange = (typeof TimeRange)[keyof typeof TimeRange];

// ==================== 資料模型 ====================

/**
 * 分析摘要
 */
export interface AnalyticsSummary {
  total_realized_pl: number;
  total_realized_pl_pct: number;
  total_cost_basis: number;
  total_sell_amount: number;
  total_sell_fee: number;
  transaction_count: number;
  currency: string;
  time_range: string;
  start_date: string;
  end_date: string;
}

/**
 * 績效資料
 */
export interface PerformanceData {
  asset_type: AssetType;
  name: string;
  realized_pl: number;
  realized_pl_pct: number;
  cost_basis: number;
  sell_amount: number;
  transaction_count: number;
}

/**
 * 最佳/最差表現資產
 */
export interface TopAsset {
  symbol: string;
  name: string;
  asset_type: AssetType;
  realized_pl: number;
  realized_pl_pct: number;
  cost_basis: number;
  sell_amount: number;
}

// ==================== 輔助函式 ====================

/**
 * 取得時間範圍的顯示名稱
 */
export function getTimeRangeLabel(timeRange: TimeRange): string {
  const labels: Record<TimeRange, string> = {
    [TimeRange.WEEK]: "本週",
    [TimeRange.MONTH]: "本月",
    [TimeRange.QUARTER]: "本季",
    [TimeRange.YEAR]: "本年",
    [TimeRange.ALL]: "全部",
  };
  return labels[timeRange];
}

/**
 * 取得所有時間範圍選項
 */
export function getTimeRangeOptions() {
  return Object.values(TimeRange).map((value) => ({
    value,
    label: getTimeRangeLabel(value),
  }));
}

/**
 * 格式化金額（加上千分位）
 */
export function formatCurrency(
  amount: number,
  currency: string = "TWD"
): string {
  return `${currency} ${amount.toLocaleString("zh-TW", {
    minimumFractionDigits: 0,
    maximumFractionDigits: 0,
  })}`;
}

/**
 * 格式化百分比
 */
export function formatPercentage(value: number): string {
  const sign = value >= 0 ? "+" : "";
  return `${sign}${value.toFixed(2)}%`;
}

/**
 * 判斷是否為正值
 */
export function isPositive(value: number): boolean {
  return value >= 0;
}

// ==================== 未實現損益分析 ====================

/**
 * 未實現損益摘要
 */
export interface UnrealizedSummary {
  total_cost: number;
  total_market_value: number;
  total_unrealized_pl: number;
  total_unrealized_pct: number;
  holding_count: number;
  currency: string;
}

/**
 * 未實現績效資料
 */
export interface UnrealizedPerformance {
  asset_type: AssetType;
  name: string;
  cost: number;
  market_value: number;
  unrealized_pl: number;
  unrealized_pct: number;
  holding_count: number;
}

/**
 * Top 未實現損益資產
 */
export interface UnrealizedTopAsset {
  symbol: string;
  name: string;
  asset_type: AssetType;
  quantity: number;
  avg_cost: number;
  current_price: number;
  cost: number;
  market_value: number;
  unrealized_pl: number;
  unrealized_pct: number;
}

// ==================== 資產配置分析 ====================

/**
 * 按資產類型的配置
 */
export interface AllocationByType {
  asset_type: AssetType;
  name: string;
  market_value: number;
  percentage: number;
  count: number;
}

/**
 * 按個別資產的配置
 */
export interface AllocationByAsset {
  symbol: string;
  name: string;
  asset_type: AssetType;
  market_value: number;
  percentage: number;
  quantity: number;
}

/**
 * 資產配置摘要
 */
export interface AllocationSummary {
  total_market_value: number;
  by_type: AllocationByType[];
  by_asset: AllocationByAsset[];
  currency: string;
  as_of_date: string;
}

// ==================== 績效趨勢分析 ====================

/**
 * 績效趨勢資料點
 */
export interface PerformanceTrendPoint {
  date: string;
  market_value: number;
  cost: number;
  unrealized_pl: number;
  unrealized_pct: number;
  realized_pl: number;
  realized_pct: number;
  total_pl: number;
  total_pct: number;
  holding_count: number;
}

/**
 * 按資產類型的績效趨勢
 */
export interface PerformanceTrendByType {
  asset_type: AssetType;
  name: string;
  data: PerformanceTrendPoint[];
}

/**
 * 績效趨勢摘要
 */
export interface PerformanceTrendSummary {
  start_date: string;
  end_date: string;
  total_data: PerformanceTrendPoint[];
  by_type: PerformanceTrendByType[];
  currency: string;
  data_point_count: number;
}

// Settings 相關型別
export interface DiscordSettings {
  webhook_url: string;
  enabled: boolean;
  report_time: string; // HH:MM 格式
  monthly_report_enabled: boolean; // 啟用月度報告
  monthly_report_day: number; // 每月發送日期 (1-10)
  yearly_report_enabled: boolean; // 啟用年度報告
  yearly_report_month: number; // 每年發送月份 (1-12)
  yearly_report_day: number; // 每年發送日期 (1-10)
}

export interface AllocationSettings {
  tw_stock: number;
  us_stock: number;
  crypto: number;
  rebalance_threshold: number;
}

export interface NotificationSettings {
  daily_billing: boolean;
  subscription_expiry: boolean;
  installment_completion: boolean;
  expiry_days: number;
}

export interface SettingsGroup {
  discord: DiscordSettings;
  allocation: AllocationSettings;
  notification: NotificationSettings;
}

export interface UpdateSettingsGroupInput {
  discord?: DiscordSettings;
  allocation?: AllocationSettings;
  notification?: NotificationSettings;
}

/**
 * 每日績效快照
 */
export interface DailyPerformanceSnapshot {
  id: string;
  snapshot_date: string;
  total_market_value: number;
  total_cost: number;
  total_unrealized_pl: number;
  total_unrealized_pct: number;
  total_realized_pl: number;
  total_realized_pct: number;
  holding_count: number;
  currency: string;
  created_at: string;
  updated_at: string;
}
